### PR TITLE
NT-1770: Perimeter X Managing Denied Requests

### PR DESCRIPTION
**Pull request by Arkariang**
# 📲 What
- Manage PerimeterX Challenged responses 
- Added User Agent to `ApiRequestInterceptor.java`
- cookie for webviews

# 🛠 How
- `PerimeterXClientType` updated
- Abstracted challenged responses related code in `PerimeterXClient`
- `ApiRequestInterceptor` now has added user agent header
- Added cookie in `KSWebView.kt`

# 👀 See
- Go to the dashboard and create a challenge
- Launch the app Take a look in logcat and look for challenged responses 
Screen Shot 2021-02-22 at 8 45 02 AM

|  |  |

# 📋 QA
- Create a challenge on the dashboard
- load the app
- dismiss the captcha scree
- one the captcha screen is presented again press the button

# Story 📖

[NT-1770](https://kickstarter.atlassian.net/browse/NT-1770)

### CODE REVIEW

File: app/src/main/java/com/kickstarter/libs/perimeterx/PerimeterXClient.kt [hunk 1]
```diff
@@ -1,19 +1,123 @@
...
+        val headers = httpHeaders()?.toMap() ?: emptyMap()
 
         headers.forEach { (key, value) ->
             builder?.addHeader(key, value)
         }
+
+        if (build.isDebug) Timber.d("$LOGTAG headers: $headers added to requestBuilder: ${builder?.toString()}")
+    }
+
+    override fun vId():String = pxManager.vid
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/libs/perimeterx/PerimeterXClient.kt:70]: Consider renaming this to `visitorId` for more clarity

File: app/src/main/java/com/kickstarter/libs/perimeterx/PerimeterXClientType.kt [hunk 2]
```diff
@@ -1,9 +1,87 @@
...
+import android.content.Context
 import okhttp3.Request
+import okhttp3.Response
+import rx.Observable
+import kotlin.collections.HashMap
 
+/**
+ * Wrapper interface to handle the implementation with the PerimeterX SDK
+ */
 interface PerimeterXClientType {
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/libs/perimeterx/PerimeterXClientType.kt:12]: Great interface! Thanks for the documentation!

File: app/src/main/java/com/kickstarter/mock/factories/PXClientFactory.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import com.kickstarter.libs.Build
+import com.kickstarter.libs.perimeterx.PerimeterXClient
+import okhttp3.Response
+import rx.Observable
+import rx.subjects.PublishSubject
+import java.util.HashMap
+
+class PXClientFactory private constructor() {
+
+    open class MockPXClient(build: Build) : PerimeterXClient(build) {
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/mock/factories/PXClientFactory.kt:13]: Why did you add this class to be inside the PXClientFactory? Should it be separated?

**Review by jgsamudio — APPROVED** [review]

**Review by leighdouglas — APPROVED** [review]
Looks great! 🎉
